Tron stacking

gemdev111 opened the pull request:
closes #197

gemdev111 committed:
Tron stacking

gemdev111 committed:
remove #IF DEBUG

gemdev111 committed:
adjusted tron stacking logic

gemdev111 committed:
Merge branch 'main'

gemdev111 committed:
merge main, minor updates after typeshare

gemdev111 committed:
Merge branch 'main'

gemdev111 committed:
updates redelgate, mnor refactroing stake service

gemdev111 committed:
Minor refactoring tron service
Added tests to calulator, when using zero fees

gemdev111 committed:
Merge branch 'main'

gemdev111 committed:
simplify balabce text for tron

gemcoder21 commented on Gem/Transfer/ViewsModels/ConfirmTransferViewModel.swift:
```diff
@@ -271,9 +271,16 @@ extension ConfirmTransferViewModel {
-            try addTransaction(transaction: transaction)
+            for data in signedData {
+                let hash = try await broadcast(data: data, options: broadcastOptions)
+                if data == signedData.last {
```
let's insert each transaction

gemcoder21 commented on Services/StakeService/Sources/StakeService.swift:
```diff
@@ -39,16 +39,17 @@ public struct StakeService: Sendable {
     
     public func getActiveValidators(assetId: AssetId) throws -> [DelegationValidator] {
         try store.getValidators(assetId: assetId)
```
should we call `store.getActiveValidators` to avoid getting unnecessary validators?

gemcoder21 commented on Packages/Signer/Sources/Chains/TronSigner.swift:
```diff
@@ -30,6 +31,113 @@ public struct TronSigner: Signable {
+            }
+        case .redelegate:
+            guard case let .vote(votes) = input.extra else {
+                fatalError("vote for redelegate should be always set")
```
```suggestion
                throw("vote for redelegate should be always set")
```

gemcoder21 commented on Packages/Signer/Sources/Chains/TronSigner.swift:
```diff
@@ -30,6 +31,113 @@ public struct TronSigner: Signable {
+            return freezeOutput.json + "___" + voteOutput.json
+        case .unstake:
+            guard case let .vote(votes) = input.extra else {
+                fatalError("vote for unstaking should be always set")
```
```suggestion
                throw("vote for unstaking should be always set")
```

gemcoder21 commented on Packages/Signer/Sources/Chains/TronSigner.swift:
```diff
@@ -30,6 +31,113 @@ public struct TronSigner: Signable {
+
+    public func signStake(input: SignerInput, privateKey: Data) throws -> String {
+        guard case let .stake(_, stakeType) = input.type else {
+            fatalError("Invalid input type for staking")
```
```suggestion
            throw("Invalid input type for staking")
```

gemcoder21 commented on Packages/Signer/Sources/Chains/TronSigner.swift:
```diff
@@ -30,6 +31,113 @@ public struct TronSigner: Signable {
+        switch stakeType {
+        case .stake:
+            guard case let .vote(votes) = input.extra else {
+                fatalError("vote for stacking should be always set")
```
```suggestion
                throw("vote for stacking should be always set")
```

gemcoder21 commented on Packages/Blockchain/Sources/Tron/TronService.swift:
```diff
@@ -312,11 +420,82 @@ extension TronService: ChainSyncable {
+
+            return DelegationBase(
+                assetId: chain.assetId,
+                state: Date() < completionDate ? .pending : .awaitingWithdrawal,
```
`Date() < completionDate ? .pending : .awaitingWithdrawal` move this to `let state`

gemcoder21 commented on Packages/Blockchain/Sources/Tron/TronService.swift:
```diff
@@ -312,11 +420,82 @@ extension TronService: ChainSyncable {
+                guard totalVotes > 0 else { return String(reward) }
+                let proportion = Double(vote.vote_count) / Double(totalVotes)
+                let rewardForVote = (Double(reward) * proportion)
+                return String(format: "%.0f", rewardForVote)
```
Can we convert to `BigInt()` and then pass to `DelegationBase()` to make sure it's always big int value

gemdev111 committed:
Merge branch 'main'

gemdev111 committed:
Merge branch 'main'

gemdev111 committed:
TronService updates coinBalance & stakeBalance with proper values
Updates based on @gemcoder21 review

gemdev111 committed:
tron adjustments

gemdev111 committed:
stake balance for tron simplify

gemdev111 committed:
Stake details adjustments
added missed apr style
added logic to hide empty validator apr
Removed tron checking in banner setup service

gemdev111 committed:
removed unused aprTextStyle

gemdev111 committed:
Merge branch 'main'

gemdev111 committed:
signer stake result type updates

gemdev111 committed:
updates
